R/oauth-cache.R

```diff
@@ -0,0 +1,138 @@
+## see oauth-cache.R in httr
+## nothing there is exported, so to make desired changes to oauth caching
+## I have to copy and mutate here
+
+gargle_cache_path <- function() {
+  "~/.R/gargle/gargle-oauth"
+}
+
+use_cache <- function(cache = getOption("gargle.oauth_cache")) {
+  if (length(cache) != 1) {
+    stop("cache should be length 1 vector", call. = FALSE)
+  }
+  if (!is.logical(cache) && !is.character(cache)) {
+    stop("Cache must either be logical or string (file path)")
+  }
+
+  # If missing, see if it's ok to use one, and cache the results of
+  # that check in a global option.
+  if (is.na(cache)) {
+    cache <- can_use_cache()
+    options("gargle.oauth_cache" = cache)
+  }
+  ## cache is now TRUE, FALSE or path
+
+  if (isFALSE(cache)) {
+    return(NULL)
+  }
+
+  if (isTRUE(cache)) {
+    cache <- gargle_cache_path()
+  }
+
+  if (!file.exists(cache)) {
+    create_cache(cache)
+  }
+  return(cache)
+}
+
+can_use_cache <- function(path = gargle_cache_path()) {
+  file.exists(path) || should_cache(path)
+}
+
+should_cache <- function(path = gargle_cache_path()) {
+  if (!interactive()) return(FALSE)
+
+  cat("Use a local file ('", path, "'), to cache OAuth access credentials ",
+      "between R sessions?\n", sep = "")
+  utils::menu(c("Yes", "No")) == 1
+}
+
+create_cache <- function(path = gargle_cache_path()) {
+
+  cache_parent <- dirname(path)
+  if (!dir.exists(cache_parent)) {
+    dir.create(cache_parent, recursive = TRUE)
+  }
+
+  file.create(path, showWarnings = FALSE)
+  if (!file.exists(path)) {
+    stop("Failed to create local cache ('", path, "')", call. = FALSE)
+  }
+
+  "!DEBUG cache exists: `path`"
+
+  # Protect cache as much as possible
+  Sys.chmod(path, "0600")
+
+  ## TODO(jennybc): this only looks in exact directory of cache path, not up
```

Comment:
To really do this right, you need a notion of project and to ask if the cache file appears to be in a project that is an R package or git repo.